### PR TITLE
[REVIEW] Add libcudf support for __contains__  

### LINKED ISSUE #2699: [FEA] A libcudf function to be used for __contains__ to find an element.
**Is your feature request related to a problem? Please describe.**
A libcudf function to be used for `__contains__` to find element. The current implantation is not efficient.

**Describe the solution you'd like**
Using `thrust::any_of`.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
[REVIEW] libcudf support for __contains__ python function has been provided. Test cases has been added to cover all the scenarios.

close #2699

### CODE REVIEW

File: cpp/src/search/search.cu [hunk 1]
```diff
@@ -129,4 +130,53 @@ gdf_column upper_bound(table const& t,
...
+            __device__ bool operator()(gdf_index_type i){
+               return compare(i, 0);
+            }
+            row_equality_comparator<nullable> compare;
+};
+
+bool contains(gdf_column const& column, gdf_scalar const& value)
+{
+  // No element to compare against
+  if (column.size == 0 || value.is_valid == false) {
```
Comment (jrhemstad) [inline cpp/src/search/search.cu]: If `value.is_valid == false`, then you should return `cudf::has_nulls(column)`, i.e., if a user is searching for a null value, then return true if any null values exist in `column`.

File: cpp/src/search/search.cu [hunk 2]
```diff
@@ -129,4 +130,53 @@ gdf_column upper_bound(table const& t,
...
+
+  gdf_column* tmp_column = const_cast<gdf_column *> (&column);
+  gdf_column* tmp_value = val.get();
+
+  // Creating a single column device table
+  auto d_t = device_table::create(1, &tmp_column, stream);
+  auto d_value = device_table::create(1, &tmp_value, stream);
+  auto data_it = thrust::make_counting_iterator(0);
+
+  if (column.null_count > 0) {
```
Comment (jrhemstad) [inline cpp/src/search/search.cu]:
```suggestion
  if (cudf::has_nulls(column)) {
```

File: cpp/include/cudf/search.hpp [hunk 3]
```diff
@@ -99,5 +99,23 @@ gdf_column upper_bound(table const& t,
                        std::vector<bool> const& desc_flags,
                        bool nulls_as_largest = true);
 
+/**---------------------------------------------------------------------------*
+ * @brief Find if the `value` is present in the `column`
+ *
```
Comment (jrhemstad) [inline cpp/include/cudf/search.hpp]: Need to document the requirement that `value` and `column` have the same `dtype`.

File: cpp/include/cudf/search.hpp [hunk 4]
```diff
@@ -99,5 +99,23 @@ gdf_column upper_bound(table const& t,
...
+ *  Single Column:
+ *      idx      0   1   2   3   4
+ *   column = { 10, 20, 20, 30, 50 }
+ *  Scalar:
+ *   value = { 20 }
+ *   result = true
+ *
+ * @param column   A gdf column
+ * @param value    A scalar value to be searched in `column`
```
Comment (jrhemstad) [inline cpp/include/cudf/search.hpp]:
```suggestion
 * @param value    A scalar value to search for in `column`
```

File: cpp/src/search/search.cu [hunk 5]
```diff
@@ -129,4 +130,53 @@ gdf_column upper_bound(table const& t,
...
+  // No element to compare against
+  if (column.size == 0 || value.is_valid == false) {
+      return false;
+  }
+
+  cudaStream_t stream = 0;
+  // Create column with scalar's data
+  gdf_column_wrapper val (1, value.dtype, gdf_dtype_extra_info{}, "");
+  RMM_TRY(RMM_ALLOC(&val.get()->data, cudf::size_of(value.dtype), stream));
+  cudaMemcpyAsync(val.get()->data, (void*) &value.data,
```
Comment (jrhemstad) [inline cpp/src/search/search.cu]:
```suggestion
  CUDA_TRY(cudaMemcpyAsync(val->data, (void*) &value.data,
```

File: cpp/src/search/search.cu [hunk 6]
```diff
@@ -129,4 +130,53 @@ gdf_column upper_bound(table const& t,
...
+};
+
+bool contains(gdf_column const& column, gdf_scalar const& value)
+{
+  // No element to compare against
+  if (column.size == 0 || value.is_valid == false) {
+      return false;
+  }
+
+  cudaStream_t stream = 0;
```
Comment (jrhemstad) [inline cpp/src/search/search.cu]: Create an API in the `detail` namespace that accepts a stream argument defaulted to `0` and make the public API simply call the `detail::contains` function.

File: cpp/include/cudf/search.hpp (hunk 3 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/search.hpp]: @jrhemstad Should I also throw an exception for this case in function body?

Comment (jrhemstad) [reply cpp/include/cudf/search.hpp]: > this supports if the column is a int and value is float and vice versa.

I'm not following. The function as it is implemented now can only support `column.dtype == scalar.dtype`.

Comment (rgsl888prabhu) [reply cpp/include/cudf/search.hpp]: Ok, I will take care of it.

**Comment by rgsl888prabhu** [comment]
@kkraus14

**Pull request merged**